cmd/crossplane/main.go

```diff
@@ -20,27 +20,39 @@ import (
 	"os"
 	"path/filepath"
 
-	"gopkg.in/alecthomas/kingpin.v2"
+	kingpin "gopkg.in/alecthomas/kingpin.v2"
 	"sigs.k8s.io/controller-runtime/pkg/client/config"
 	"sigs.k8s.io/controller-runtime/pkg/manager"
 	runtimelog "sigs.k8s.io/controller-runtime/pkg/runtime/log"
 	"sigs.k8s.io/controller-runtime/pkg/runtime/signals"
 
 	"github.com/crossplaneio/crossplane/pkg/apis"
 	"github.com/crossplaneio/crossplane/pkg/controller"
+	extensionsController "github.com/crossplaneio/crossplane/pkg/controller/extensions"
+	"github.com/crossplaneio/crossplane/pkg/extensions"
 	"github.com/crossplaneio/crossplane/pkg/logging"
 )
 
 func main() {
 	var (
 		log = logging.Logger
 
+		// top level app definition
 		app        = kingpin.New(filepath.Base(os.Args[0]), "An open source multicloud control plane.").DefaultEnvars()
 		debug      = app.Flag("debug", "Run with debug logging.").Short('d').Bool()
 		syncPeriod = app.Flag("sync", "Controller manager sync period duration such as 300ms, 1.5h or 2h45m").
 				Short('s').Default("1h").Duration()
+
+		// default crossplane command and args
+		crossplaneCmd = app.Command(filepath.Base(os.Args[0]), "An open source multicloud control plane.").Default()
+
+		// extensions commands and args
+		extCmd       = app.Command("extension", "Perform operations on extensions")
```

Comment:
I am not sure I fully get these commands.  When would I want to use one over another?

Reply:
Good question, there may be better ways to do this in the future.  These aren't necessarily commands that a human would run, they are invoked by the controllers/jobs.  Here are more thorough explanations about their functionality and usage:

* Extension Manager deployment/controller: invokes the `crossplane extension manage` command which registers and starts up the extension controllers, controller manager, etc. and waits for events on the Extension CRD types (e.g. `ExtensionRequest`).
* Extension Manager package unpacking job: invokes the `crossplane extension unpack` command which will unpack and parse an extension package's filesystems and translate them to install artifacts such as CRDs and an `Extension` record.

One future possibility would be to create a new binary for this EM functionality instead of including them in the main crossplane binary right now.  It's definitely much easier to include everything in one binary right now, but we can consider moving it out in the future.

Another possibility that `spf13/cobra` supports, not sure if `kingpin` supports it, is to make these "hidden" commands, so they don't show up in the help/usage.  Since they aren't intended to be invoked by humans, that could be an option to look into.

Also increasing the documentation here to explain these commands better would be wise, I'll take a note to do that.  Any other questions or feedback about these that I can address?

Reply:
Is the intent for `ExtensionManager` run as a separate (from the `Crossplane`) deployment/pod/process. If so, what's is rational? If not, why not to run multiple controllers under one `crossplane` umbrella (as we already do).

Reply:
Yes, the Extension Manager runs as a completely separate deployment/pod from the main Crossplane pod.  I chose this because the EM needs cluster-admin perms so that it can create arbitrary RBAC perms for arbitrary extensions that it installs (similar to Tiller), and I wanted to scope that elevated permissions level to as small a surface area as possible.  The regular Crossplane deployment/pod still runs with its scoped permissions.

Reply:
Here's the new separate `ClusterRoleBinding` for the extension manager: https://github.com/crossplaneio/crossplane/pull/444/files#diff-fd191f2469ca86cc5e39de007c9e828cR1

And the new deployment that uses the new `extension manage` subcommand: https://github.com/crossplaneio/crossplane/pull/444/files#diff-bc876258b3f300036e11032cf60f42a6R27

Reply:
Have you considered to place them into separate `cmd` entries? For example:
```
cmd
├── crossplane
│   └── main.go
├── extension-manager
│   └── main.go
├── extension-operations
│   └── main.go
└── extension-[un]packer
    └── main.go
```

I think this will create a cleaner and more intuitive implementation.

Reply:
or into:

Reply:
good feedback @ichekrygin, i will try to refactor to clean up the main entry points as suggested. i've not done that before with kingpin, but i've done it with cobra before. hope there are no quirks.